system/modals/GlobalModalManagerV2.tsx

```diff
@@ -0,0 +1,138 @@
+import styles from './GlobalModalManagerV2.module.scss';
+
+import * as React from 'react';
+
+export interface ModalProviderProps {
+  children?: React.ReactNode;
+}
+
+export interface ModalComponentV2<P = {}> extends React.FC<P & CommonModalProps> {
+  unmountDelayMS?: number;
+}
+
+export interface ModalState {
+  key: string;
+  component: ModalComponentV2;
+  props: object;
+}
+
+export interface CommonModalProps {
+  isClosing: boolean;
+  close: () => void;
+}
+
+export interface ModalContextTypeV2 {
+  activeModal: ModalState | null;
+  closingModals: { [key: string]: ModalState };
+  showModal: <P>(key: string, component: ModalComponentV2<P> | null, props?: P) => void;
+  hideCurrentModal: () => void;
+  hideModal: (key: string) => void;
+}
+
+const defaultModalContext: ModalContextTypeV2 = {
+  activeModal: null,
+  closingModals: {},
+  showModal: () => {},
+  hideCurrentModal: () => {},
+  hideModal: () => {},
+};
+export const ModalContextV2 = React.createContext(defaultModalContext);
+
+function newModalState(): ModalState | null {
+  return null;
+}
+
+/** 
+ * NOTE(@elijaharita): Provides a context that allows any of its descendants to
+ * control modals via `useModalV2()`. Must be paired with `<ModalsV2 />` in
+ * order for modals to show up. 
+ * */
```

Comment:
could you explain what the global modal manager could be used for? is it to hide and show certain information throughout the website?

Reply:
@APiligrim i added some more info about modals. does this answer your question?

Reply:
The global modal manager as a concept already exists in the template. @APiligrim this is an upgrade to the system that already exists.

You can "pass" a component to the Global Modal Manager and render the DOM element precisely where you want it in the DOM tree, so that the application performs better and more predictably, which is important for high quality applications.

Modals are generally an after thought for developers, developers implement modal systems that produce something we call "technical debt". We want to produce a system that every Modal can use so that we don't have redundancy in code, or difficult to maintain code. It also allows us to add features and control the behavior of the UI with a lot more finesse.